### PR TITLE
allow non-loggedin users to access public bucket

### PR DESCRIPTION
**Pull request by kmkanagaraj**
## Description
- Hide make bucket action for non-loggedin user
- Show upload file action if the bucket/prefix is writable
- show bucket search only for loggedin users
- don't require login to access  public bucket

## Motivation and Context
Re-write browser code to follow best practices and guidelines of React
#5409 #5410

## How Has This Been Tested?

Unit tests and manual

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#cacec0b4-bf49-46ac-be16-e07c6f39c41a) |
| mint-dist-xl.sh | :x: [more...](#6c5d399d-3b4c-4af9-af33-764f4311b8b0) |
<a name="cacec0b4-bf49-46ac-be16-e07c6f39c41a"></a>
### 5570-a3d53d5/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30166
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 595951e9b3a7:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 49 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a8e42e4a5c41"
  },
  "duration": "40283",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```
<a name="6c5d399d-3b4c-4af9-af33-764f4311b8b0"></a>
### 5570-a3d53d5/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30758
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 5c58373f55dd:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-t422j4vavk9w9x",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 325,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30758/aws-sdk-go-test-t422j4vavk9w9x: dial tcp 192.168.122.45:30758: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#2f1d46a9-142f-4e51-8092-66fbecddde9a) |
| mint-dist-xl.sh | :x: [more...](#47194760-7caf-41dc-9ade-4e0017b9e5bb) |
<a name="2f1d46a9-142f-4e51-8092-66fbecddde9a"></a>
### 5570-9603bef/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31800
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 34dcb02a6cce:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 48 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a8e775c9d4de"
  },
  "duration": "40181",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```
<a name="47194760-7caf-41dc-9ade-4e0017b9e5bb"></a>
### 5570-9603bef/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31863
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 6d23d354c50f:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-6k2z5a1toyhgbi",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 438,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:31863/aws-sdk-go-test-6k2z5a1toyhgbi: dial tcp 192.168.122.45:31863: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

**Comment by kaankabalak** [comment]
When I create a bucket named "login", I get stuck on the app unable to click anywhere and the extension in the address bar flickers between /minio/minio and /minio/login for a certain amount of time. I know we can name a S3 bucket as "login", however it might be a good time to discuss this issue once again to see how we can prevent these types of situations from happening.

**Comment by kmkanagaraj** [comment]
@kaankabalak yes that's a problem if we create a bucket with name 'login'. It conflicts with the login route. 

I have fixed the `minio/minio` issue. Now it flickers between `minio/` and `minio/login/`.

Other alternative would be to use hash based routing, let me check that out.

**Comment by kaankabalak** [comment]
I found the earlier discussion about the issue where the user sets "login" as bucket name, maybe it might be helpful for us to see how we could approach this.

https://github.com/minio/minio/issues/4766
https://github.com/minio/minio/issues/4809

File: browser/app/js/App.js [hunk 1]
```diff
@@ -34,13 +34,15 @@ const AuthorizedRoute = ({ component: Component, ...rest }) => (
...
 
-export const App = ({ match }) => (
-  <Switch>
-    <AuthorizedRoute exact path={match.url} component={Browser} />
-    <Route path={`${match.url}/login`} component={Login} />
-    <AuthorizedRoute path={`${match.url}/:bucket/*`} component={Browser} />
-    <AuthorizedRoute path={`${match.url}/:bucket`} component={Browser} />
-  </Switch>
-)
+export const App = ({ match }) => {
```
Comment (kaankabalak) [inline browser/app/js/App.js]: Is there a particular reason why we have `match` here? As far as I can see we don't use it anywhere apart from the unit test

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#0b222396-cec2-49ff-9210-8f3b9489cc25) |
| mint-dist-xl.sh | :x: [more...](#68bdc2b7-02e0-4fc3-8b3a-9b1c70b90b6d) |
<a name="68bdc2b7-02e0-4fc3-8b3a-9b1c70b90b6d"></a>
### 5570-11ac43f/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32106
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 6817807f697a:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-o3iivda1cd00bq",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 312,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:32106/aws-sdk-go-test-o3iivda1cd00bq: dial tcp 192.168.122.45:32106: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```
<a name="0b222396-cec2-49ff-9210-8f3b9489cc25"></a>
### 5570-11ac43f/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31989
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp d7c33096268b:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 0 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 48 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a8fbf712a8d8"
  },
  "duration": "40332",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```

File: browser/app/js/App.js (hunk 1 above)
Comment (kmkanagaraj) [reply browser/app/js/App.js]: Not being used, will remove it

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#d5c205a0-8c46-4fdb-8f57-f73cfa07c830) |
| mint-dist-xl.sh | :x: [more...](#15f87047-32f4-4a52-8b57-ae53e562d596) |
<a name="d5c205a0-8c46-4fdb-8f57-f73cfa07c830"></a>
### 5570-48303bf/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32002
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp ba6fe9544599:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 0 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 47 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a8fc839bd47b"
  },
  "duration": "40115",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```
<a name="15f87047-32f4-4a52-8b57-ae53e562d596"></a>
### 5570-48303bf/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:30788
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 7736acdb88cb:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-b5zoovfu4ayojc",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 351,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:30788/aws-sdk-go-test-b5zoovfu4ayojc: dial tcp 192.168.122.45:30788: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```

File: browser/app/js/browser/Login.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ export class Login extends React.Component {
         password: document.getElementById("secretKey").value
       })
       .then(res => {
-        history.push(minioBrowserPrefix)
+        history.push("/")
```
Comment (kaankabalak) [inline browser/app/js/browser/Login.js:48]: Should we remove the `import { minioBrowserPrefix } from "../constants"` line on top of this file as we don't use it here anymore?

Comment (kmkanagaraj) [reply browser/app/js/browser/Login.js]: yes, will do

**Comment by balamurugana** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#575d0191-87a8-47dc-987d-c347eae5794c) |
| mint-dist-xl.sh | :x: [more...](#73838e7a-4c7a-40c9-a469-1957b3f78780) |
<a name="73838e7a-4c7a-40c9-a469-1957b3f78780"></a>
### 5570-b26c9a0/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:32511
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp eebc4033d4e1:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-okmbvv9uat0jop",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 291,
  "error": "RequestError: send request failed\ncaused by: Put http://minikube:32511/aws-sdk-go-test-okmbvv9uat0jop: dial tcp 192.168.122.45:32511: getsockopt: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 12 tests successfully.
```
<a name="575d0191-87a8-47dc-987d-c347eae5794c"></a>
### 5570-b26c9a0/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT: minikube:31889
ACCESS_KEY:      minio
SECRET_KEY:      ***REDACTED***
ENABLE_HTTPS:    0
SERVER_REGION:   us-east-1
MINT_DATA_DIR:   /mint/data
MINT_MODE:       core

To get logs, run 'docker cp 2e087367b856:/mint/log /tmp/mint-logs'
(1/12) Running aws-sdk-go tests ... done in 1 seconds
(2/12) Running aws-sdk-php tests ... FAILED in 49 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5a90ce7a895b3"
  },
  "duration": "40061",
  "status": "FAIL",
  "error": "Expected getBucketPolicy to fail with 404"
}

Executed 1 out of 12 tests successfully.
```

**Pull request merged**